elasticsearch-transport/spec/elasticsearch/connections/collection_spec.rb

```diff
@@ -0,0 +1,251 @@
+
+  describe '#add' do
+
+    let(:collection) do
+      described_class.new(connections: [ double('connection', host: 'A', dead?: false),
+                                         double('connection', host: 'B', dead?: false) ])
+    end
+
+    context 'when an array is provided' do
+
+      before do
+        collection.add([double('connection', host: 'C', dead?: false),
+                        double('connection', host: 'D', dead?: false)])
+      end
+
+      it 'adds the connections' do
+        expect(collection.size).to eq(4)
+      end
+    end
+
+    context 'when an element is provided' do
+
+      before do
+        collection.add(double('connection', host: 'C', dead?: false))
+      end
+
+      it 'adds the connection' do
+        expect(collection.size).to eq(3)
+      end
+    end
+  end
+
+  describe '#remove' do
+
+    let(:connections) do
+      [ double('connection', host: 'A', dead?: false),
+        double('connection', host: 'B', dead?: false) ]
+    end
+
+    let(:collection) do
+      described_class.new(connections: connections)
+    end
+
+    context 'when an array is provided' do
+
+      before do
+        collection.remove(connections)
+      end
+
+      it 'removes the connections' do
+        expect(collection.size).to eq(0)
+      end
+    end
+
+    context 'when an element is provided' do
+
+      let(:connections) do
+        [ double('connection', host: 'A', dead?: false),
+          double('connection', host: 'B', dead?: false) ]
+      end
+
+      before do
+        collection.remove(connections.first)
+      end
+
+      it 'removes the connection' do
+        expect(collection.size).to eq(1)
+      end
+    end
+  end
+
+  describe '#get_connection' do
+
+    context 'when all connections are dead' do
+
+      let(:connection_a) do
+        Elasticsearch::Transport::Transport::Connections::Connection.new(host: { host: 'A' })
+      end
+
+      let(:connection_b) do
+        Elasticsearch::Transport::Transport::Connections::Connection.new(host: { host: 'B' })
+      end
+
+      let(:collection) do
+        described_class.new(connections: [connection_a, connection_b])
+      end
+
+      before do
+        connection_a.dead!.dead!
+        connection_b.dead!
+      end
+
+      it 'returns the connection with the least failures' do
+        expect(collection.get_connection.host[:host]).to eq('B')
+      end
+    end
+
+    context 'when multiple threads are used' do
+
+      let(:connections) do
+        100.times.collect do |i|
+          Elasticsearch::Transport::Transport::Connections::Connection.new(host: { host: i })
+        end
+      end
+
+      let(:collection) do
+        described_class.new(connections: connections)
+      end
+
+      it 'allows threads to select connections in parallel' do
+        threads = []
+        threads << Thread.new do
+          2000.times do |i|
+            collection.get_connection
+          end
+        end
+        threads.collect {|t| t.join}
+      end
```

Comment:
This code is not creating connections in parallel.  2000.times should be on outer loop 
2000.times do |i|
  threads  << Thread.new do 
    collection.get_connection
end
end

Reply:
Thanks, that was a silly mistake.